pandas/core/indexes/base.py

```diff
@@ -6591,6 +6591,9 @@ def _maybe_cast_listlike_indexer(self, target) -> Index:
         """
         Analogue to maybe_cast_indexer for get_indexer instead of get_loc.
         """
+        if not hasattr(target, "dtype") and self.dtype == object:
+            # Avoid inference for object since we are casting back later anyway
+            return Index(target, dtype=self.dtype)
```

Comment:
hmm this might prevent us from doing intentional inference in _maybe_downcast_for_indexing

Reply:
Looking at `_maybe_downcast_for_indexing` more closely, we are downcasting object dtype index to the other dtype, because this can give a performance improvement, right?

This might actually also be true for object/string combo, but I think correctness if more important, and if missing values are handled differently by downcasting, we should avoid that for now for this combo, until we can solve that better.

I will just make the check a bit more specific for object dtype then (even more ugly, but at least fixing this issue)